File: small_problems/medium_2/ex3.rb
LOWER_CASE = ('a'..'z')
UPPER_CASE = ('A'..'Z')

def letter_percentages(str)
  lower_count = 0
  upper_count = 0
  neither_count = 0
  str.each_char do |char|
    if LOWER_CASE.include?(char) then lower_count += 1
    elsif UPPER_CASE.include?(char) then upper_count += 1
    else neither_count += 1
    end
  end
  lower_percent = (lower_count.to_f / str.length) * 100
  upper_percent = (upper_count.to_f / str.length) * 100
  neither_percent = (neither_count.to_f / str.length) * 100
  { lowercase: lower_percent, uppercase: upper_percent, neither: neither_percent }
end

p letter_percentages('abCdef 123') == { lowercase: 50.0, uppercase: 10.0, neither: 40.0 }
p letter_percentages('AbCd +Ef') == { lowercase: 37.5, uppercase: 37.5, neither: 25.0 }
p letter_percentages('123') == { lowercase: 0.0, uppercase: 0.0, neither: 100.0 }

File: js_small_problems/string_processing/ex3.js
function letterCaseCount(string) {
  let lowercaseMatches = string.match(/[a-z]/g) || [];
  let uppercaseMatches = string.match(/[A-Z]/g) || [];
  let lowercase = lowercaseMatches.length;
  let uppercase = uppercaseMatches.length;
  let neither = string.length - lowercase - uppercase;

  console.log({
    lowercase,
    uppercase,
    neither,
  });
}

letterCaseCount('abCdef 123');  // { lowercase: 5, uppercase: 1, neither: 4 }
letterCaseCount('AbCd +Ef');    // { lowercase: 3, uppercase: 3, neither: 2 }
letterCaseCount('123');         // { lowercase: 0, uppercase: 0, neither: 3 }
letterCaseCount('');            // { lowercase: 0, uppercase: 0, neither: 0 }